src/transformers/adapter_bart.py

```diff
@@ -0,0 +1,305 @@
+        self.config = parent.config
+
+    @property
+    def adapter_config_key(self):
+        return "mh_adapter"
+
+    @property
+    def layer_norm(self):
+        # if layer norm was applied before attention/ FFW block, don't use it here
+        if self.config.normalize_before:
+            return None
+        else:
+            return self.parent.self_attn_layer_norm
+
+
+class BartCrossAttentionAdaptersModule(BartAdaptersBaseMixin, nn.Module):
+    def __init__(self, parent):
+        super().__init__()
+        # keep a reference to the parent module without registering as a submodule
+        object.__setattr__(self, "parent", parent)
+        self.config = parent.config
+
+    @property
+    def adapter_config_key(self):
+        return "cross_adapter"
+
+    @property
+    def layer_norm(self):
+        # if layer norm was applied before attention/ FFW block, don't use it here
+        if self.config.normalize_before:
+            return None
+        else:
+            return self.parent.encoder_attn_layer_norm
+
+
+class BartOutputAdaptersModule(BartAdaptersBaseMixin, nn.Module):
+    def __init__(self, parent):
+        super().__init__()
+        # keep a reference to the parent module without registering as a submodule
+        object.__setattr__(self, "parent", parent)
+        self.config = parent.config
+
+    @property
+    def adapter_config_key(self):
+        return "output_adapter"
+
+    @property
+    def layer_norm(self):
+        # if layer norm was applied before attention/ FFW block, don't use it here
+        if self.config.normalize_before:
+            return None
+        else:
+            return self.parent.final_layer_norm
+
+
+class BartEncoderLayerAdaptersMixin:
+    """Adds adapters to the BartEncoderLayer or BartDecoderLayer module of BART."""
+
+    def _init_adapter_modules(self):
+        self.attention_adapters = BartSelfAttentionAdaptersModule(self)
+        self.output_adapters = BartOutputAdaptersModule(self)
+        self.attention_adapters._init_adapter_modules()
+        self.output_adapters._init_adapter_modules()
+
+    def add_fusion_layer(self, adapter_names):
+        self.attention_adapters.add_fusion_layer(adapter_names)
+        self.output_adapters.add_fusion_layer(adapter_names)
+
+    def add_adapter(self, adapter_name: str, layer_idx: int):
+        self.attention_adapters.add_adapter(adapter_name, layer_idx)
+        self.output_adapters.add_adapter(adapter_name, layer_idx)
+
+    def enable_adapters(self, adapter_names: list, unfreeze_adapters: bool, unfreeze_attention: bool):
+        self.attention_adapters.enable_adapters(adapter_names, unfreeze_adapters, unfreeze_attention)
+        self.output_adapters.enable_adapters(adapter_names, unfreeze_adapters, unfreeze_attention)
+
+
+class BartDecoderLayerAdaptersMixin(BartEncoderLayerAdaptersMixin):
+    """Adds adapters to the BartEncoderLayer or BartDecoderLayer module of BART."""
```

Comment:
Doesn't this only add adapters to the BartEncoderLayer?

Reply:
Of course, thanks for pointing that out. 👍